lib/util.js

```diff
@@ -748,27 +748,74 @@ var util = {
   /**
    * @api private
    */
-  addPromisesToRequests: function addPromisesToRequests(constructor, PromiseDependency) {
-    PromiseDependency = PromiseDependency || null;
-    if (!PromiseDependency && typeof Promise !== 'undefined') {
-      PromiseDependency = Promise;
-    }
-    if (typeof PromiseDependency !== 'function') {
-      delete constructor.prototype.promise;
-      return;
-    }
-    constructor.prototype.promise = function promise() {
-      var self = this;
-      return new PromiseDependency(function(resolve, reject) {
-        self.on('complete', function(resp) {
-          if (resp.error) {
-            reject(resp.error);
-          } else {
-            resolve(resp.data);
-          }
+  addPromises: function addPromises(constructors, PromiseDependency) {
+    if (!AWS) AWS = require('./core');
+    if (!AWS.config) require('./config');
+    if (PromiseDependency === undefined) PromiseDependency = AWS.config.getPromisesDependency();
+    if (typeof PromiseDependency !== 'function') var deletePromise = true;
+    if (!Array.isArray(constructors)) constructors = [constructors];
+
+    var promisifyMethod = function(methodName) {
+      return function promise() {
+        var self = this;
+        return new PromiseDependency(function(resolve, reject) {
+          self[methodName](function(err, data) {
+            if (err) {
+              reject(err);
+            } else {
+              resolve(data);
+            }
+          });
         });
-        self.runTo();
-      });
+      };
+    };
+
+    for (var ind = 0; ind < constructors.length; ind++) {
+      var constructor = constructors[ind];
+      switch (constructor.name) {
```

Comment:
It looks like `constructor.name` may not work in all the browsers we support and could have issues when minifiers are used:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/name#Browser_compatibility

Can you do an equality check against the constructor instead?